### ISSUE TITLE
Umbrel App

### ISSUE THREAD

**Issue by bijeebuss** [fedimint/fedimint#4242]
Is anybody working on or planning to start working on an Umbrel App soon? If not I would like to start on that. Are there any glaring concerns with doing that at this point?

**Comment by EthnTuttle** [comment]
> I remember @EthnTuttle doing something with StartOS and @Kodylow has experience with dockerizing Fedimint. @douglaz you have the most operational experience, any concerns with beginning packaging?

Yup! https://github.com/EthnTuttle/fedimintd-startos

I have successfully packaged up fedimintd to run on StartOS and am (slowly) making some UI (https://github.com/fedimint/ui) changes to accomodate the packaging. If you're working on bundling for Umbrel, please join us in the Discord (chat.fedimint.org) if you want to chat about it. Also, feel free to tag me here.

In addition to Umbrel, I have considered packaging for AMI/Terraform/Ansible, as this seems helpful for devs/admins. There are a couple more "nodes in a box" that might be a fit too: My Citadel, raspi-blitz, My Node.

Glad someone else is thinking of this!

**Comment by Kodylow** [comment]
gonna take a look at this tonight, hop in the discord and we'll move forward on it

**Comment by Kodylow** [comment]
Couple things: 
1. generally the fedimint guardian should not also be running a gateway, running both on the same machine removes the separation of concerns that's fundamental to fedimint's custody model.
2. the follower/leader distinction only applies for the setup process, after the setup all the nodes are just equal guardians and  any user facing app using the federation will attempt to get threshold responses from the guardians (depending on the method) so all the guardians would have to be online with good uptime
3. Highly recommend you prioritize and first go with the lightning gateway: that's a light side service to the lightning node that generally will have infrequent comms to end users when they attempt lightning payments, vs running a guardian node is an entirely separate service than the lightning node and will generally have longer and more frequent comms with every member of the fedimint

**Comment by elsirion** [comment]
The client requires in many cases to talk to t out of n (e.g. 3 of 4) guardians to be sure they aren't being lied to.

**Comment by Kodylow** [comment]
Update here after working with bijeebuss for a bit:

1. We require a couple ports to be openly accessible for clients to connect to for both gateway and guardian nodes. Umbrel doesn't expose publicly ports normally and requires you to use Tor hidden services, @elsirion what's the status on gateway/guardian support over Tor have we tested that? Think I remember you had something like last year for Tor?

2. Going to try to get the gateway working first because it only requires the 1 public port for clients to connect to plus umbrel already runs an LND lightning node so makes things nice and easy (unless there's other umbrel apps using the grpc interceptor since it's only got the one hook in? they probably have a check for that on lightning terminal, should see how that works)

3. After gateway working do guardian, but we should do some testing around it since running both over Tor + on little RasPis will be different than what we normally cover and use for guardians

**Comment by douglaz** [comment]
> In addition to Umbrel, I have considered packaging for AMI/Terraform/Ansible, as this seems helpful for devs/admins. There are a couple more "nodes in a box" that might be a fit too: My Citadel, raspi-blitz, My Node.
> 
> Glad someone else is thinking of this!

FYI, I'm starting to test things in kubernetes, which is widely used in enterprise setups (but it could also be used in home environments)

**Comment by douglaz** [comment]
> @douglaz you have the most operational experience, any concerns with beginning packaging?

I'm not sure fedimint is mature enough for these kind of setups. Ideally umbrel apps should work out of box with minimal configuration. We definitely should find some way of automatically exposing the guardians to the internet (perhaps using some third-party service).

**Comment by bijeebuss** [comment]
I think tor would also solve the port issue and would work as long as the client apps support connecting over tor. However, I'm not a big fan of tor for anything that needs to work reliably so I think tunneling is the way to go

**Comment by douglaz** [comment]
> 1. Exposing ports: There is a cloudflare tunnel app in the umbrel app store that can achieve this. The only issue is that the FM_API_URL and FM_P2P_URL parameters need to be set as environment variables before starting fedimintd. The way umbrel apps work is you provide a docker compose file and an environment variables file and together they define the service. Umbrel doesn't provide a way to set these variables on a per user basis so either we could A: update fedimintd and/or the UI to be able to change these variables without having to set them in the compose file. B: See if we can create a PR to umbrel to add a way to set initialization parameters when installing an app.


Note that these values can be arguments to fedimintd on command line and if you can figure out some way of pulling these configurations from somewhere they can be set up somewhat dynamically. For instance on k8s I'm testing something like:
```bash
            - sh
            - -c
            - fedimintd --api-url ws://${POD_IP}:80 --p2p-url fedimint://${POD_IP}:8173
```

but this value may also come from another binary, so you could think of something like:
```bash
fedimintd  --api-url $(magical-tunnel get-or-create-tls --ws-address) --p2p-url fedimint://$(magical-tunnel get-or-create-tcp --address) 
```

**Comment by bijeebuss** [comment]
> but this value may also come from another binary, so you could think of something like:
> 
> ```shell
> fedimintd --api-url $(magical-tunnel get-or-create-tls --to localhost:80 --ws-address) --p2p-url fedimint://$(magical-tunnel get-or-create-tcp --to localhost:8173 --address) 
> ```

Ok thanks this is giving me some ideas. Ill get back to you on this

**Comment by bijeebuss** [comment]
> dev call: P2P ports over TOR will be hard to do (will slow down consensus), API has MVP that @oleonardolima is looking into (useful for Umbrel). But having different API connections going through different circuits is harder.

I agree, I think the cloudflare tunnel app will do the job there. For setting the parameters I like this idea proposed [here](https://github.com/Radiokot/umbrel-cloudflared/issues/6):
- Create a configuration service with a UI to make all the preparations; The service writes the parameters to a .env file
- Add this configuration service to the compose with custom healthcheck for it, so it reports healthy status only if successfully configured;
- Make the main service dependent on the configuration with depends_on.

I started working on the configuration service with a simple web app. I think with this approach I can configure it so that I don't need to create another image (besides the config UI) that builds on top of fedimint's images.

**Comment by elsirion** [comment]
@bijeebuss if you are still interested, we recently added the necessary features to support node packages and have a reference integration for Start9: #7117

Right now no UI client supports the new Iroh networking, so you'd need to use `fedimint-cli` for testing.

**Comment by bijeebuss** [comment]
Thanks, I haven't had time to pursue this.  If anybody want to take over, go ahead

**Comment by m1sterc001guy** [comment]
This has been solved: https://apps.umbrel.com/app/fedimintd
